protocols/gossipsub/src/metrics.rs

```diff
@@ -0,0 +1,255 @@
+    /// determined by users on the network.  This limit permits a fixed amount of topics to allow,
+    /// in-addition to the mesh topics.
+    pub max_never_subscribed_topics: usize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Config {
+            max_topics: DEFAULT_MAX_EXPLICITLY_SUBSCRIBED_TOPICS,
+            max_never_subscribed_topics: DEFAULT_MAX_NEVER_SUBSCRIBED_TOPICS,
+        }
+    }
+}
+
+/// Whether we have ever been subscribed to this topic.
+type EverSubscribed = bool;
+
+/// Reasons why a peer was included in the mesh.
+#[derive(PartialEq, Eq, Hash, Encode, Clone)]
+pub enum Inclusion {
+    /// Peer was a fanaout peer.
+    Fanaout,
+    /// Included from random selection.
+    Random,
+    /// Peer subscribed.
+    Subscribed,
+    /// Peer was included to fill the outbound quota.
+    Outbound,
+}
+
+/// Reasons why a peer was removed from the mesh.
+#[derive(PartialEq, Eq, Hash, Encode, Clone)]
+pub enum Churn {
+    /// Peer disconnected.
+    Dc,
+    /// Peer had a bad score.
+    BadScore,
+    /// Peer sent a PRUNE.
+    Prune,
+    /// Peer unsubscribed.
+    Unsub,
+    /// Too many peers.
+    Excess,
+}
+
+/// A collection of metrics used throughout the Gossipsub behaviour.
+pub struct Metrics {
+    /* Configuration parameters */
+    /// Maximum number of topics for which we store metrics. This helps keep the metrics bounded.
+    max_topics: usize,
+    /// Maximum number of topics for which we store metrics, where the topic in not one to which we
+    /// have subscribed at some point. This helps keep the metrics bounded, since these topics come
+    /// from received messages and not explicit application subscriptions.
+    max_never_subscribed_topics: usize,
+
+    /* Auxiliary variables */
+    /// Information needed to decide if a topic is allowed or not.
+    topic_info: HashMap<TopicHash, EverSubscribed>,
+
+    /* Metrics per known topic */
+    /// Status of our subscription to this topic. This metric allows analyzing other topic metrics
+    /// filtered by our current subscription status.
+    topic_subscription_status: Family<TopicHash, Gauge>,
+    /// Number of peers subscribed to each topic. This allows us to analyze a topic's behaviour
+    /// regardless of our subscription status.
+    topic_peers_count: Family<TopicHash, Gauge>,
+
+    /* Metrics regarding mesh state */
+    /// Number of peers in our mesh. This metric should be updated with the count of peers for a
+    /// topic in the mesh regardless of inclusion and churn events.
+    mesh_peer_counts: Family<TopicHash, Gauge>,
+    /// Number of times we include peers in a topic mesh for different reasons.
+    mesh_peer_inclusion_events: Family<(TopicHash, Inclusion), Counter>,
+    /// Number of times we remove peers in a topic mesh for different reasons.
+    mesh_peer_churn_events: Family<(TopicHash, Churn), Counter>,
```

Comment:
`(TopicHash, Churn)` would encode as `mesh_peer_churn_events_total{<some-topic-hash>="<some-churn-variant>"} <some-value>`. I don't think that is what you actually want, right? Instead I would create a new struct like:

```rust
#[derive(Encode)]
struct MeshPeerChurnEventsLabels {
    hash: TopicHash,
    churn: Churn,
}
```

And then here do:

```suggestion
    mesh_peer_churn_events: Family<MeshPeerChurnEventsLabels, Counter>,
```

Reply:
Thanks Max, I've fixed it as suggested